packages/babel-helper-module-transforms/src/rewrite-live-references.js

```diff
@@ -306,4 +306,41 @@ const rewriteReferencesVisitor = {
       }
     },
   },
+  "ForOfStatement|ForInStatement"(path) {
+    const { scope, node } = path;
+    const { left } = node;
+    const { exported, scope: programScope } = this;
+
+    // if it's a variable declaration, such as for (let {foo} of []) {}
+    // then no transformation is needed
+    if (!t.isVariableDeclaration(left)) {
+      let didTransform = false;
+      const bodyPath = path.get("body");
+      const loopBodyScope = bodyPath.scope;
+      for (const name of Object.keys(t.getOuterBindingIdentifiers(left))) {
+        if (
+          exported.get(name) &&
+          programScope.getBinding(name) === scope.getBinding(name)
+        ) {
+          didTransform = true;
+          if (loopBodyScope.hasOwnBinding(name)) {
+            loopBodyScope.rename(name);
+          }
+        }
+      }
+      if (!didTransform) {
+        return;
+      }
+      const newLoopId = scope.generateUidIdentifier();
```

Comment:
Nit: we can pass an "hint" to generate a better name

```suggestion
      const newLoopId = scope.generateUidBasedOnNode(left);
```

Reply:
Done!